src/app/hooks/useViewTracker/index.jsx

```diff
@@ -0,0 +1,87 @@
+import { useEffect, useState, useContext, useRef } from 'react';
+import { useInView } from 'react-intersection-observer';
+// Polyfill IntersectionObsesrver, e.g. for IE11
+import 'intersection-observer';
+
+import { ServiceContext } from '#contexts/ServiceContext';
+import { RequestContext } from '#contexts/RequestContext';
+import { buildATIClickParams } from '#containers/ATIAnalytics/params';
+import { getComponentInfo } from '#app/lib/analyticsUtils/index';
+import { sendEventBeacon } from '#containers/ATIAnalytics/beacon';
+
+const VIEWED_DURATION_MS = 1000;
+
+const useViewTracker = ({ pageData, componentName, actionLabel } = {}) => {
+  let eventTrackingProps;
+  const requestContext = useContext(RequestContext);
+  const serviceContext = useContext(ServiceContext);
+  const timer = useRef(null);
+  const [viewSent, setViewSent] = useState(false);
+  const [ref, inView] = useInView({
+    threshold: 0.5,
+  });
+
+  try {
+    eventTrackingProps = buildATIClickParams(
+      pageData,
+      requestContext,
+      serviceContext,
+    );
+  } catch (error) {
```

Comment:
should we swallow the error here? If we don't know of any specific errors that we can expect it might better to let the error propogate so we can debug any issues in live effectively or at least log the error out.

Reply:
My thinking here was that the problem with these functions are they are passed complex objects that need to be in a very specific structure for every page type. if they are not they will break the app. I think it's bad practice to log stuff to the browser console - I think we should work on clientside logging instead. Also, I think we would notice something is wrong from a glance in ATI dashboard if this hook stopped working. I'd remove the try catches around these functions if they accepted only a few primitive values which I think could be all they need but that requires a bit of a refactor o these functions and the ATI business logic in general.

Reply:
Thanks your thoughts here 👍 I'd say in the absence of clientside logging, I think we should log to the console if any error occurs here otherwise we would have no visibility of any errors in live. I'd say it's bad practice to log to console routinely, but I'm not so sure in an error case? I think we'd notice on the ATI dashboard in a situation where we had entirely broken view tracking but if we only broke it in a smaller set of scenarios (e.g. because of some specific missing data in `pageData` on one page type) we would have no visibility of errors in that smaller scenario. Finally, cypress would pick-up any stray console errors during our cypress testing, without logging to console we wouldn't be able to take advantage of this addtional coverage.

I do agree with the try catches though based on your reasoning above. I won't not approve the PR based on this, just giving my opinion 🙂

Reply:
ah ok that's a good point about Cypress. I just didn't really see the point of logging to the browser console in the off-chance someone notices the message in amongst all of the other noise. I'll log to console then. Does Cypress pick console.warn and console.error? Maybe should log the former?

Reply:
I think it catches unhandled errors, `console.error` might fall into that bracket but not certain, here's the docs on cypress mechanism: https://docs.cypress.io/api/events/catalog-of-events#Uncaught-Exceptions

I imagine even if cypress doesn't log them the errors would be useful whilst developing the tracking mechanisms and shouldn't happen for users if our solution is stable. Again, just an opinion, I get the argument on noise 👍